### PR TITLE
style: Avoid looping through every selector more than twice.

### PR DESCRIPTION
**Pull request by emilio**
I've left the Invalidation stuff on its own since that's more complex, but I
think this may help a bit (perhaps not too much though?) with the slow rebuild
times.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17939)

### CODE REVIEW

**Comment by emilio** [comment]
r? @bzbarsky 

I have a try run ongoing, so maybe wait until it finishes, but assuming I haven't messed up...

**Comment by emilio** [comment]
(or @bholley, fwiw)

**Comment by emilio** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=819ba72325479b92908dd72ba581e451ff925c8d&selectedJob=120154348

File: components/style/stylist.rs [hunk 1]
```diff
@@ -1524,86 +1465,108 @@ impl SelectorMapEntry for RevalidationSelectorAndHashes {
...
-/// whose matching depends on the selector up in the chain.
-///
-/// TODO(emilio): We can optimize when matching only siblings to only match the
-/// rightmost selector until a descendant combinator is found, I guess, and in
-/// general when we're sharing at depth `n`, to the `n + 1` sequences of
-/// descendant combinators.
-///
-/// I don't think that in presence of the bloom filter it's worth it, though.
-struct RevalidationVisitor;
+struct StylistSelectorVisitor<'a> {
```
Comment (bzbarsky) [inline components/style/stylist.rs]: This needs documentation.  Please add some!

File: components/style/stylist.rs [hunk 2]
```diff
@@ -1524,86 +1465,108 @@ impl SelectorMapEntry for RevalidationSelectorAndHashes {
...
+        _: SelectorIter<SelectorImpl>,
+        combinator: Option<Combinator>
+    ) -> bool {
+        self.needs_revalidation =
+            self.needs_revalidation || combinator.map_or(false, |c| c.is_sibling());
+
+        // NOTE(emilio): This works properly right now because we can't store
+        // complex selectors in nested selectors, otherwise we may need to
+        // rethink this.
+        self.passed_rightmost_selector =
```
Comment (bzbarsky) [inline components/style/stylist.rs]: I think the comment about how this is called before we visit any simple selectors in the non-rightmost ComplexSelector should somehow get put in here.  That's pretty important for the way we use this member.  I know you reference this again below in the comments on Component::ID.

File: components/style/stylist.rs [hunk 3]
```diff
@@ -1524,86 +1465,108 @@ impl SelectorMapEntry for RevalidationSelectorAndHashes {
...
+    passed_rightmost_selector: bool,
+    mapped_ids: &'a mut BloomFilter,
+    attribute_dependencies: &'a mut BloomFilter,
+    style_attribute_dependency: &'a mut bool,
+    state_dependencies: &'a mut ElementState,
+}
 
-impl SelectorVisitor for RevalidationVisitor {
-    type Impl = SelectorImpl;
+fn component_needs_revalidation(c: &Component<SelectorImpl>) -> bool {
```
Comment (bzbarsky) [inline components/style/stylist.rs]: I am so glad the weird double-negative thing of the old setup has gone away.  I always had to detangle it very carefully.

**Review by bzbarsky — APPROVED** [review]
Looks good with the nits.

File: components/style/stylist.rs [hunk 4]
```diff
@@ -1524,86 +1465,108 @@ impl SelectorMapEntry for RevalidationSelectorAndHashes {
...
+    match *c {
+        Component::AttributeInNoNamespaceExists { .. } |
+        Component::AttributeInNoNamespace { .. } |
+        Component::AttributeOther(_) |
+        Component::Empty |
+        // FIXME(bz) We really only want to do this for some cases of id
+        // selectors.  See
+        // https://bugzilla.mozilla.org/show_bug.cgi?id=1369611
```
Comment (emilio) [inline components/style/stylist.rs]: Will do as a followup.

**Comment by jdm** [comment]
```
error[E0432]: unresolved import `style::stylist::needs_revalidation`
  --> /home/servo/buildbot/slave/linux-dev/build/tests/unit/style/stylist.rs:23:5
   |
23 | use style::stylist::needs_revalidation;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `needs_revalidation` in `stylist`

error: aborting due to previous error
```

**Comment by emilio** [comment]
@bors-servo r=bzbarsky

 * Added a test-only function to be able to test the revalidation stuff.

**Pull request merged**